netlogo-gui/src/main/app/CodeTabsPanel.scala

```diff
@@ -0,0 +1,125 @@
+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo
+package org.nlogo.app
+
+import java.awt.event.{ MouseAdapter, MouseEvent, WindowAdapter, WindowEvent }
+
+import javax.swing.JTabbedPane
+import javax.swing.event.{ ChangeEvent, ChangeListener }
+
+import scala.collection.mutable
+
+import org.nlogo.app.codetab.{ ExternalFileManager, MainCodeTab, TemporaryCodeTab }
+import org.nlogo.app.common.{ Events => AppEvents }
+import org.nlogo.core.I18N
+import org.nlogo.app.interfacetab.InterfaceTab
+import org.nlogo.window.GUIWorkspace
+import org.nlogo.window.Event
+import org.nlogo.window.Event.LinkParent
+import org.nlogo.window.Events._
+
+// When a separate code tab window is created, an instance of this class owns the CodeTabs.
+// When there is no separate code tab window, no such instance exists, and all
+// CodeTabs belong to Tabs. [ Thinking of Tabs as AppTabsPanel makes the parallels between
+// CodeTabsPanel and Tabs clearer - both are JTabbedPanes that contain and manage tabs.]
+// CodeTabsPanel and Tabs are both instances of AbstractTabsPanel, which implements their shared behavior. AAB 10/2020
+
+class CodeTabsPanel(workspace:             GUIWorkspace,
+                    interfaceTab:          InterfaceTab,
+                    externalFileManager:   ExternalFileManager,
+                    val mainCodeTab:       MainCodeTab,
+                    val externalFileTabs:  mutable.Set[TemporaryCodeTab])
+  extends AbstractTabsPanel(workspace, interfaceTab, externalFileManager)
+  with ChangeListener
+  with AfterLoadEvent.Handler
+  with LinkParent
+  with org.nlogo.window.LinkRoot {
+
+  locally {
+    addChangeListener(this)
+  }
+
+  // frame is the App's AppFrame, (treated as a java.awt.Frame) AAB 10/2020
+  val frame = workspace.getFrame
+
+  // CodeTabContainer contains the CodeTabsPanel and is owned by frame
+  // It is currently implemented as a JDialog. AAB 10/2020
+  val codeTabContainer = new CodeTabContainer(frame, this)
+  val codeTabsPanel = this
+
+  def asAbstractTabsPanel = { this.asInstanceOf[AbstractTabsPanel] }
```

Comment:
I don't understand the use for this.  Isn't `CodeTabsPanel` already an `AbstractTabsPanel` since it `extends` it?  Why do we need to cast it?  Why do we need a method to cast it if we do?